src/network/networkd-dhcp6.c

```diff
@@ -35,6 +36,93 @@ static bool dhcp6_get_prefix_delegation(Link *link) {
                       RADV_PREFIX_DELEGATION_BOTH);
 }
 
+static bool dhcp6_has_preferred_subnet_id(Link *link) {
+        assert(link->network);
+        return link->network->router_prefix_subnet_id > -1;
+}
+
+static int dhcp6_get_preferred_delegated_prefix(
+                Manager* manager,
+                Link *link,
+                const struct in6_addr *pd_prefix,
+                uint8_t pd_prefix_len,
+                struct in6_addr *ret_addr) {
+        int r;
+        union in_addr_union prefix;
+        union in_addr_union pd_prefix_union;
+        const uint8_t prefix_bits = 64 - pd_prefix_len;
+        const uint64_t n_prefixes = UINT64_C(1) << prefix_bits;
+        const int64_t subnet_id = link->network->router_prefix_subnet_id;
+        _cleanup_free_ char *assigned_buf = NULL;
+
+        pd_prefix_union.in6 = *pd_prefix;
+
+        /* We start off with the original PD prefix we have been assigned and
+         * iterate from there */
+        prefix.in6 = *pd_prefix;
+        log_link_debug(link, "subnet id %" PRIi64, subnet_id);
+
+        if (subnet_id > -1) {
+                /* If the link has a preference for a particular subnet id try to allocate that */
+                if ((uint64_t)subnet_id >= n_prefixes) {
+                        return log_link_debug_errno(link,
+                                        SYNTHETIC_ERRNO(ERANGE),
+                                        "subnet id %" PRIi64 " is out of range. Only have %" PRIu64 " subnets",
+                                         subnet_id,
+                                         n_prefixes);
+                } else {
+                        r = in_addr_prefix_nth(AF_INET6, &prefix, 64, subnet_id);
+                        if (r < 0)
+                                return log_link_debug_errno(link,
+                                                SYNTHETIC_ERRNO(ERANGE),
+                                                "subnet id %" PRIi64 " is out of range. Only have %" PRIu64 " subnets",
+                                                subnet_id,
+                                                n_prefixes);
+
+                        /* Verify that the prefix we did calculate fits in the pd prefix. */
+                        r = in_addr_prefix_covers(AF_INET6, &pd_prefix_union, pd_prefix_len, &prefix);
+                        assert(r == 1); /* This should never happen as we checked the prefix size beforehand */
```

Comment:
we have `assert_se()` for assert checks with side effects that shall never be optimized away. That allows you to do this in one expression:

```c
assert_se(in_addr_prefix_covers(…) > 0);
```